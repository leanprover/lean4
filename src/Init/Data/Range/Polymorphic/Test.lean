/-
Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Paul Reichert
-/
module

prelude
public import Init.Core
import Init.Data.Order.Lemmas2

public section

open Std

instance : Total (α := Nat) (· ≤ ·) := inferInstance
