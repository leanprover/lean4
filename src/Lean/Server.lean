/-
Copyright (c) 2020 Marc Huisinga. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.

Authors: Marc Huisinga, Wojciech Nawrocki
-/
import Lean.Server.Watchdog
import Lean.Server.FileWorker
import Lean.Server.Rpc
