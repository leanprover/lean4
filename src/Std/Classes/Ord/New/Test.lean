import Std.Classes.Ord.New.PartiallyComparable
import Std.Data.TreeMap

open Std

example (α : Type u) [Ord α] [pc : PartiallyComparable α] [LawfulPartiallyComparableOrd pc]
    [LawfulLinearOrder pc] : Unit := Id.run do
  let mut m : DTreeMap α (fun _ => Nat) := ∅
  let x := m.get sorry sorry
  sorry
