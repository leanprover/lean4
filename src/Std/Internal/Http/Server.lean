/-
Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Sofia Rodrigues
-/
module

prelude
public import Std.Internal.Async
public import Std.Internal.Async.TCP
public import Std.Sync.CancellationToken
public import Std.Internal.Http.Server.Config
public import Std.Internal.Http.Server.Connection

public section

/-!
# HTTP Server

This module defines a simple, asynchronous HTTP/1.1 server implementation.

It provides the `Std.Http.Server` structure, which encapsulates all server state, and functions for
starting, managing, and gracefully shutting down the server.

The server runs entirely using `Async` and uses a shared `CancellationToken` to signal shutdowns.
Each active client connection is tracked, and the server automatically resolves its shutdown
promise once all connections have closed.
-/

namespace Std.Http

open Std.Internal.IO.Async TCP

set_option linter.all true

/--
The `Server` structure holds all state required to manage the lifecycle of an HTTP server, including
connection tracking and shutdown coordination.
-/
structure Server where

  /--
  The cancellation token used for shutting down all connections and the server.
  -/
  cancellation : Std.CancellationToken

  /--
  Active HTTP connections
  -/
  activeConnections : Std.Mutex UInt64

  /--
  Indicates when the server has successfully shutdown
  -/
  shutdownPromise : IO.Promise Unit

  /--
  Configuration of the server
  -/
  config : Std.Http.Config

namespace Server

/--
Create a new `Server` structure with an optional configuration.
-/
def new (config : Std.Http.Config := {}) : IO Server := do
  let cancellation ← Std.CancellationToken.new
  let activeConnections ← Std.Mutex.new 0
  let shutdownPromise : IO.Promise Unit ← IO.Promise.new

  return { cancellation, activeConnections, shutdownPromise, config }

/--
Triggers the cancellation of all the requests and the accept loop in the server.
-/
@[inline]
def shutdown (s : Server) : Async Unit :=
  s.cancellation.cancel

/--
Waits for the server to shutdown.
-/
@[inline]
def waitShutdown (s : Server) : Async Unit := do
  Async.ofAsyncTask (s.shutdownPromise.result?.map (Option.map Except.ok · |>.getD (.error "promise dropped at server shutdown task")))

/--
Triggers the cancellation of all the requests and the accept loop in the server and wait for the server
to be cancelled.
-/
@[inline]
def shutdownAndWait (s : Server) : Async Unit := do
  s.cancellation.cancel
  s.waitShutdown

@[inline]
private def frameCancellation (s : Server) (action : Async α) : Async α := do
  s.activeConnections.atomically (modify (· + 1))

  let result ← action

  s.activeConnections.atomically do
    modify (· - 1)

    if (← get) = 0 ∧ (← s.cancellation.isCancelled) then
      s.shutdownPromise.resolve ()

  return result

/--
Start a new HTTP/1.1 server on the given socket address. This function uses the `Async` for handling
tasks and TCP connections, it also returns a `Server` structure that can be used to the cancellation
of the server.
-/
def serve
    (addr : Net.SocketAddress)
    (onRequest : Request Body → Async (Response Body))
    (config : Config := {}) (backlog : UInt32 := 128) : Async Server := do

  let httpServer ← Server.new config

  let server ← Socket.Server.mk
  server.bind addr
  server.listen backlog

  background do
    frameCancellation httpServer do
      while not (← httpServer.cancellation.isCancelled) do
        let client ← server.accept
        background (frameCancellation httpServer (serveConnection client onRequest config) )

  return httpServer

end Std.Http.Server
