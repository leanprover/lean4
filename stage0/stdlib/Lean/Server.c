// Lean compiler output
// Module: Lean.Server
// Imports: Init Lean.Server.Watchdog Lean.Server.FileWorker Lean.Server.Rpc Lean.Server.CodeActions
#include <lean/lean.h>
#if defined(__clang__)
#pragma clang diagnostic ignored "-Wunused-parameter"
#pragma clang diagnostic ignored "-Wunused-label"
#elif defined(__GNUC__) && !defined(__CLANG__)
#pragma GCC diagnostic ignored "-Wunused-parameter"
#pragma GCC diagnostic ignored "-Wunused-label"
#pragma GCC diagnostic ignored "-Wunused-but-set-variable"
#endif
#ifdef __cplusplus
extern "C" {
#endif
lean_object* initialize_Init(uint8_t builtin, lean_object*);
lean_object* initialize_Lean_Server_Watchdog(uint8_t builtin, lean_object*);
lean_object* initialize_Lean_Server_FileWorker(uint8_t builtin, lean_object*);
lean_object* initialize_Lean_Server_Rpc(uint8_t builtin, lean_object*);
lean_object* initialize_Lean_Server_CodeActions(uint8_t builtin, lean_object*);
static bool _G_initialized = false;
LEAN_EXPORT lean_object* initialize_Lean_Server(uint8_t builtin, lean_object* w) {
lean_object * res;
if (_G_initialized) return lean_io_result_mk_ok(lean_box(0));
_G_initialized = true;
res = initialize_Init(builtin, lean_io_mk_world());
if (lean_io_result_is_error(res)) return res;
lean_dec_ref(res);
res = initialize_Lean_Server_Watchdog(builtin, lean_io_mk_world());
if (lean_io_result_is_error(res)) return res;
lean_dec_ref(res);
res = initialize_Lean_Server_FileWorker(builtin, lean_io_mk_world());
if (lean_io_result_is_error(res)) return res;
lean_dec_ref(res);
res = initialize_Lean_Server_Rpc(builtin, lean_io_mk_world());
if (lean_io_result_is_error(res)) return res;
lean_dec_ref(res);
res = initialize_Lean_Server_CodeActions(builtin, lean_io_mk_world());
if (lean_io_result_is_error(res)) return res;
lean_dec_ref(res);
return lean_io_result_mk_ok(lean_box(0));
}
#ifdef __cplusplus
}
#endif
