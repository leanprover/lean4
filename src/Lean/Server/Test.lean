/-
Copyright (c) 2025 Lean FRO. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.

Authors: Sebastian Ullrich
-/
prelude
import Lean.Server.Test.Cancel
import Lean.Server.Test.Runner
