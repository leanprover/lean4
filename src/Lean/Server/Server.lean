import Lean.Environment
import Lean.Elab.Frontend
import Lean.Data.Lsp
import Lean.Data.Json.FromToJson
import Init.System.IO

namespace Lean.Server

open IO
open Lean
open Lean.JsonRpc
open Lean.Lsp
open Lean.Elab

-- LSP indexes text with rows and colums
abbrev DocumentText := Array String

structure Document :=
(version : Int)
(text : DocumentText)
(headerEnv : Environment)

abbrev DocumentMap := RBMap DocumentUri Document (fun a b => Decidable.decide (a < b))

structure ServerState :=
(i o : FS.Handle)
(openDocumentsRef : IO.Ref DocumentMap)

def parseParams (paramType : Type*) [HasFromJson paramType] (params : Json) : IO paramType :=
match @fromJson? paramType _ params with
| some parsed => pure parsed
| none        => throw (userError "got param with wrong structure")

def runFrontend (text : String) : IO (Environment × Environment × MessageLog) := do
let inputCtx := Parser.mkInputContext text "<input>";
envNul ← mkEmptyEnvironment;
match Parser.parseHeader envNul inputCtx with
| (header, parserState, messages) => do
  (env, messages) ← processHeader header messages inputCtx;
  parserStateRef ← IO.mkRef parserState;
  cmdStateRef ← IO.mkRef $ Command.mkState env messages {};
  IO.processCommands inputCtx parserStateRef cmdStateRef;
  cmdState ← cmdStateRef.get;
  pure (env, cmdState.env, cmdState.messages)

def updateFrontend (env : Environment) (input : String) : IO (Environment × MessageLog) := do
let inputCtx := Parser.mkInputContext input "<input>";
parserStateRef ← IO.mkRef ({} : Parser.ModuleParserState);
cmdStateRef    ← IO.mkRef $ Command.mkState env;
IO.processCommands inputCtx parserStateRef cmdStateRef;
cmdState ← cmdStateRef.get;
pure (cmdState.env, cmdState.messages)

def msgToDiagnostic (text : DocumentText) (m : Lean.Message) : Diagnostic :=
-- Lean Message line numbers are 1-based while LSP Positions are 0-based.
let lowLn := m.pos.line - 1;
let low : Lsp.Position := ⟨lowLn, (text.get! lowLn).codepointPosToUtf16Pos m.pos.column⟩;
let high : Lsp.Position := match m.endPos with
| some endPos =>
  let highLn := endPos.line - 1;
  ⟨highLn, (text.get! highLn).codepointPosToUtf16Pos endPos.column⟩
| none        => low;
let range : Range := ⟨low, high⟩;
let severity := match m.severity with
| MessageSeverity.information => DiagnosticSeverity.information
| MessageSeverity.warning     => DiagnosticSeverity.warning
| MessageSeverity.error       => DiagnosticSeverity.error;
let source := "Lean 4 server";
let message := toString (format m.data);
{ range := range
, severity? := severity
, source? := source
, message := message}

namespace ServerState

def findOpenDocument (s : ServerState) (key : DocumentUri) : IO Document := do
openDocuments ← s.openDocumentsRef.get;
match openDocuments.find? key with
| some doc => pure doc
| none     => throw (userError "got unknown document uri")

def updateOpenDocuments (s : ServerState) (key : DocumentUri) (val : Document) : IO Unit :=
s.openDocumentsRef.modify (fun documents => (documents.erase key).insert key val)

-- Clears diagnostics for the document version 'd'.
-- TODO how to clear all diagnostics? Sending version 'none' doesn't seem to work
def clearDiagnostics (s : ServerState) (uri : DocumentUri) (d : Document) : IO Unit :=
writeLspNotification s.o "textDocument/publishDiagnostics"
  { uri := uri
  , version? := d.version
  , diagnostics := #[] : PublishDiagnosticsParams }

def sendDiagnostics (s : ServerState) (uri : DocumentUri) (d : Document) (log : MessageLog) : IO Unit :=
let diagnostics := log.msgs.map (msgToDiagnostic d.text);
writeLspNotification s.o "textDocument/publishDiagnostics"
  { uri := uri
  , version? := d.version
  , diagnostics := diagnostics.toArray : PublishDiagnosticsParams }

def handleDidOpen (s : ServerState) (p : DidOpenTextDocumentParams) : IO Unit := do
let doc := p.textDocument;
let text := doc.text.splitOnEOLs;
(headerEnv, env, msgLog) ← runFrontend ("\n".intercalate text);
let newDoc : Document := ⟨doc.version, text.toArray, headerEnv⟩;
s.openDocumentsRef.modify (fun openDocuments => openDocuments.insert doc.uri newDoc);
s.sendDiagnostics doc.uri newDoc msgLog

def handleDidChange (s : ServerState) (p : DidChangeTextDocumentParams) : IO Unit := do
let docId := p.textDocument;
let changes := p.contentChanges;
oldDoc ← s.findOpenDocument docId.uri;
some newVersion ← pure docId.version? | throw (userError "expected version number");
if newVersion <= oldDoc.version then do
  throw (userError "got outdated version number")
else changes.forM $ fun change =>
  match change with
  | TextDocumentContentChangeEvent.rangeChange (range : Range) (newText : String) => do
    let newDocText := replaceRange oldDoc.text range newText;
    (newEnv, msgLog) ← updateFrontend oldDoc.headerEnv ("\n".intercalate newDocText.toList);
    let newDoc : Document := ⟨newVersion, newDocText, oldDoc.headerEnv⟩;
    s.updateOpenDocuments docId.uri newDoc;
    -- Clients don't have to clear diagnostics, so we clear them
    -- for the *previous* version here.
    s.clearDiagnostics docId.uri oldDoc;
    s.sendDiagnostics docId.uri newDoc msgLog

  | TextDocumentContentChangeEvent.fullChange (text : String) =>
    throw (userError "got content change that replaces the full document (not supported)")

def handleDidClose (s : ServerState) (p : DidCloseTextDocumentParams) : IO Unit := do
-- TODO is any extra cleanup needed?
s.openDocumentsRef.modify (fun openDocuments => openDocuments.erase p.textDocument.uri)

def handleNotification (s : ServerState) (method : String) (params : Json) : IO Unit := do
let h := (fun paramType [HasFromJson paramType] (handler : ServerState → paramType → IO Unit) =>
  parseParams paramType params >>= handler s);
match method with
| "textDocument/didOpen"   => h DidOpenTextDocumentParams handleDidOpen
| "textDocument/didChange" => h DidChangeTextDocumentParams handleDidChange
| "textDocument/didClose"  => h DidCloseTextDocumentParams handleDidClose
| _                        => throw (userError "got unsupported notification method")

def handleRequest (s : ServerState) (id : RequestID) (method : String) (params : Json)
  : IO Unit := do
  match method with
  | _ => throw (userError "Not supporting requests for now!")

partial def mainLoop : ServerState → IO Unit
| s => do
  m ← readLspMessage s.i;
  match m with
  | Message.request id method (some params) => do
    s.handleRequest id method (toJson params);
    mainLoop s
  | Message.request id "shutdown" none =>
    writeLspResponse s.o id (Json.null)
  | Message.requestNotification method (some params) => do
    s.handleNotification method (toJson params);
    mainLoop s
  | _ => throw (userError "got invalid jsonrpc message")

end ServerState

def initialize (i o : FS.Handle) : IO Unit := do
-- ignore InitializeParams for MWE
r ← readLspRequestAs i "initialize" InitializeParams;
writeLspResponse o r.id ({ capabilities := mkLeanServerCapabilities
                         , serverInfo? := some { name := "Lean 4 server"
                                               , version? := "0.0.1" }} : InitializeResult);
_ ← readLspRequestNotificationAs i "initialized" Initialized;
openDocumentsRef ← IO.mkRef (RBMap.empty : DocumentMap);
ServerState.mainLoop ⟨i, o, openDocumentsRef⟩;
Message.requestNotification "exit" none ← readLspMessage i
  | throw (userError "Expected an Exit Notification.");
pure ()

end Lean.Server

def main (n : List String) : IO UInt32 := do
i ← IO.stdin;
o ← IO.stdout;
catch (Lean.Server.initialize i o) (fun err => o.putStrLn (toString err));
pure 0
