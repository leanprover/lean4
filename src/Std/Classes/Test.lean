prelude
import Init.Core
import Init.Data.Ord
import Lean

namespace Std.Classes.Test

class BLE (α : Type u) where
  LE : α → α → Bool

class LE (α : Type u) where
  LE : α → α → Prop

class LawfulBLE (α : Type u) [LE α] [BLE α] where
  ble_iff_le : ∀ a b : α, BLE.LE a b ↔ LE.LE a b

abbrev DecidableLE (α : Type u) [LE α] :=
  DecidableRel (fun a b : α => LE.LE a b)

instance (α : Type u) [BLE α] [LE α] [LawfulBLE α] : DecidableLE α :=
  fun a b => if h : BLE.LE a b then
      .isTrue (LawfulBLE.ble_iff_le _ _ |>.mp h)
    else
      .isFalse (h.imp (LawfulBLE.ble_iff_le _ _).mpr)

class BLT (α : Type u) where
  LT : α → α → Bool

class LT (α : Type u) where
  LT : α → α → Prop

class LawfulBLT (α : Type u) [LT α] [BLT α] where
  blt_iff_lt : ∀ a b : α, BLT.LT a b ↔ LT.LT a b

abbrev DecidableLT (α : Type u) [LT α] := DecidableRel (fun a b : α => LT.LT a b)

instance (α : Type u) [BLT α] [LT α] [LawfulBLT α] : DecidableLT α :=
  fun a b => if h : BLT.LT a b then
      .isTrue (LawfulBLT.blt_iff_lt _ _ |>.mp h)
    else
      .isFalse (h.imp (LawfulBLT.blt_iff_lt _ _).mpr)

class Ord (α : Type u) where
  compare : α → α → Ordering

class Noncomputable (α : Type u) where
  Predicate : α → Prop
  nonempty : Nonempty (Subtype Predicate)
  subsingleton : Subsingleton (Subtype Predicate)

def Noncomputable.ofComputable {α : Type u} (a : α) : Noncomputable α where
  Predicate := (· = a)
  nonempty := ⟨⟨a, rfl⟩⟩
  subsingleton := .intro fun a b => Subtype.ext (a.2.trans b.2.symm)

noncomputable def Noncomputable.get (i : Noncomputable α := by infer_instance) :=
  haveI := i.nonempty
  (Classical.ofNonempty (α := Subtype i.Predicate)).val

abbrev LogicalOrd (α : Type u) := Noncomputable (Ord α)

noncomputable def LogicalOrd.instOrd {α : Type u} [i : LogicalOrd α] : Ord α :=
  Noncomputable.get

noncomputable def LogicalOrd.compare {α : Type u} [LogicalOrd α] (a b : α) : Ordering :=
  LogicalOrd.instOrd.compare a b

class OrdForOrd (α : Type u) [LogicalOrd α] [i : Ord α] where
  eq : LogicalOrd.instOrd = i

class LogicalOrd.LawfulLT (α : Type u) [LT α] [LogicalOrd α] where
  compare_eq_lt_iff_lt {a b : α} : LogicalOrd.compare a b = .lt ↔ LT.LT a b
  compare_eq_gt_iff_gt {a b : α} : LogicalOrd.compare a b = .gt ↔ LT.LT b a

class LogicalOrd.LawfulLE (α : Type u) [LE α] [LogicalOrd α] where
  compare_isLE_iff_le {a b : α} : (LogicalOrd.compare a b).isLE ↔ LE.LE a b
  compare_eq_ge_iff_ge {a b : α} : (LogicalOrd.compare a b).isLE ↔ LE.LE b a

class LogicalOrd.LawfulBLT (α : Type u) [BLT α] [LogicalOrd α] where
  compare_eq_lt_iff_lt {a b : α} : LogicalOrd.compare a b = .lt ↔ BLT.LT a b
  compare_eq_gt_iff_gt {a b : α} : LogicalOrd.compare a b = .gt ↔ BLT.LT b a

class LogicalOrd.LawfulBLE (α : Type u) [BLE α] [LogicalOrd α] where
  compare_isBLE_iff_le {a b : α} : (LogicalOrd.compare a b).isLE ↔ BLE.LE a b
  compare_eq_ge_iff_ge {a b : α} : (LogicalOrd.compare a b).isLE ↔ BLE.LE b a

class Min (α : Type u) where
  min : α → α → α

abbrev LogicalMin (α : Type u) := Noncomputable (Min α)

class Max (α : Type u) where
  max : α → α → α

abbrev LogicalMax (α : Type u) := Noncomputable (Max α)

class LogicalMin.LawfulMin (α : Type u) [LogicalMin α] [i : Min α] where
  eq : Noncomputable.get = i

class LogicalMax.LawfulMax (α : Type u) [LogicalMax α] [i : Max α] where
  eq : Noncomputable.get = i

def Ord.ofBLE {α} [BLE α] : Ord α where
  compare a b := if BLE.LE a b then
      if BLE.LE b a then .eq else .lt
    else
      .gt

noncomputable def BLE.ofLE {α} [i : Std.Classes.Test.LE α] : BLE α where
  LE a b := open scoped Classical in decide (i.LE a b)

def LogicalOrd.ofBLE {α} [BLE α] : LogicalOrd α := .ofComputable .ofBLE

class Assign {α : Type u} (a : outParam α) (b : α) where
  eq : a = b

instance (α : Type u) (a : α) : Assign a a where
  eq := rfl

attribute [local instance] LogicalOrd.ofBLE in
example [BLE α] [BLT α] [LogicalOrd.ofBLE.LawfulBLE α] [LogicalOrd.ofBLE.LawfulBLT α] (a b : α)
    (h : BLT.LT a b) : BLE.LE a b := by
  simp only [← LogicalOrd.LawfulBLE.compare_isBLE_iff_le,
    ← LogicalOrd.LawfulBLT.compare_eq_lt_iff_lt] at ⊢ h
  simp [h]

theorem s [BLE α] [BLT α] {x : LogicalOrd α} [Assign x LogicalOrd.ofBLE] [LogicalOrd.LawfulBLE α] [LogicalOrd.LawfulBLT α] {a b : α}
    (h : BLT.LT a b) : BLE.LE a b := by
  simp only [← LogicalOrd.LawfulBLE.compare_isBLE_iff_le,
    ← LogicalOrd.LawfulBLT.compare_eq_lt_iff_lt] at ⊢ h
  simp [h]

example [BLE α] [BLT α] [LogicalOrd.ofBLE.LawfulBLE α] [LogicalOrd.ofBLE.LawfulBLT α]
    (a b : α) (h : BLT.LT a b) : BLE.LE a b := s h

theorem t [BLE α] [BLT α] {x : LogicalOrd α} [LogicalOrd.LawfulBLE α] [LogicalOrd.LawfulBLT α] {a b : α}
    (h : BLT.LT a b) (_x : x = LogicalOrd.ofBLE := by rfl) : BLE.LE a b := by
  simp only [← LogicalOrd.LawfulBLE.compare_isBLE_iff_le,
    ← LogicalOrd.LawfulBLT.compare_eq_lt_iff_lt] at ⊢ h
  simp [h]

example [BLE α] [BLT α] [LogicalOrd.ofBLE.LawfulBLE α] [LogicalOrd.ofBLE.LawfulBLT α]
    (a b : α) (h : BLT.LT a b) : BLE.LE a b := t h

end Std.Classes.Test
