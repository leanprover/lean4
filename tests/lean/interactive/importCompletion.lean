import Lean.Server.Test
                     --^ textDocument/completion
import Lean.Server.Test.
                      --^ textDocument/completion
