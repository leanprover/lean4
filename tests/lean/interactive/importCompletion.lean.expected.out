{"textDocument": {"uri": "file:///importCompletion.lean"},
 "position": {"line": 0, "character": 23}}
{"items":
 [{"label": "Test.Cancel",
   "data": ["«external:file:///importCompletion.lean»", 0, 23]},
  {"label": "Test.Runner",
   "data": ["«external:file:///importCompletion.lean»", 0, 23]},
  {"label": "Test",
   "data": ["«external:file:///importCompletion.lean»", 0, 23]},
  {"label": "Test.Refs",
   "data": ["«external:file:///importCompletion.lean»", 0, 23]}],
 "isIncomplete": false}
{"textDocument": {"uri": "file:///importCompletion.lean"},
 "position": {"line": 2, "character": 24}}
{"items":
 [{"label": "Runner",
   "data": ["«external:file:///importCompletion.lean»", 2, 24]},
  {"label": "Cancel",
   "data": ["«external:file:///importCompletion.lean»", 2, 24]},
  {"label": "Refs",
   "data": ["«external:file:///importCompletion.lean»", 2, 24]}],
 "isIncomplete": false}
